### PR TITLE
(refactor): delegate generic predicate checking to @sindresorhus/is

### LINKED ISSUE #1764: Refactor code to use @sindresorhus/is
https://github.com/sindresorhus/is

Good to replace some of our own similar functions with this, e.g. some `isObject` calls etc

### PR DESCRIPTION
**Pull request by connorbrathwaite**
substituted as many generic type validations with `@sindresorhus/is`. there are edge cases that failed testing which have remain untouched.

have also seen `(foo.length)` predicates throughout the codebase which could potentially be handled with `is.arrayLike`, if that makes sense will update the pr with additional commits.  

all feedback is welcome!

Closes #1764

### CODE REVIEW

**Comment by rarkins** [comment]
@connorbrathwaite thanks for this PR. I looked into is.arrayLike and I am not sure it will work in place of our .length checks. We often find ourselves doing `it config.foo && config.foo.length)` and it would be nice to replace it but `arrayLike` returns true even if the length is empty. It would be great if they had `is.lengthyArray` returning true only if length is positive but I didn’t see something similar.

File: package.json [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "node": ">= 8.11.0"
   },
   "dependencies": {
+    "@sindresorhus/is": "^0.9.0",
```
Comment (rarkins) [inline package.json]: Let’s pin this

File: lib/workers/repository/onboarding/branch/config.js [hunk 2]
```diff
@@ -25,7 +26,7 @@ async function getOnboardingConfig(config) {
         String(branchName).replace('greenkeeper', 'renovate'),
       ];
     }
-    if (Array.isArray(ignore) && ignore.length) {
+    if (is.array(ignore)) {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/config.js]: Shouldn’t we retain length checks here?

File: lib/config/migrate-validate.js [hunk 3]
```diff
@@ -24,10 +25,10 @@ async function migrateAndValidate(config, input) {
       massagedConfig
     );
     // istanbul ignore if
-    if (warnings && warnings.length) {
+    if (is.arrayLike(warnings)) {
```
Comment (rarkins) [inline lib/config/migrate-validate.js]: Replace with `!is.empty(warnings)`

File: lib/config/migrate-validate.js [hunk 4]
```diff
@@ -24,10 +25,10 @@ async function migrateAndValidate(config, input) {
...
       logger.info({ warnings }, 'Found renovate config warnings');
     }
-    if (errors && errors.length) {
+    if (is.arrayLike(errors)) {
```
Comment (rarkins) [inline lib/config/migrate-validate.js]: Replace with `!is.empty(errors)`

File: lib/platform/github/index.js [hunk 5]
```diff
@@ -688,7 +689,7 @@ async function addReviewers(prNo, reviewers) {
 async function addLabels(issueNo, labels) {
   logger.debug(`Adding labels ${labels} to #${issueNo}`);
   const repository = config.parentRepo || config.repository;
-  if (Array.isArray(labels) && labels.length) {
+  if (is.array(labels) && labels.length) {
```
Comment (rarkins) [inline lib/platform/github/index.js]: Replace with `!is.empty(labels)`

File: lib/workers/branch/commit.js [hunk 6]
```diff
@@ -6,7 +8,7 @@ async function commitFilesToBranch(config) {
   const updatedFiles = config.updatedPackageFiles.concat(
     config.updatedLockFiles
   );
-  if (updatedFiles && updatedFiles.length) {
+  if (is.array(updatedFiles) && updatedFiles.length) {
```
Comment (rarkins) [inline lib/workers/branch/commit.js]: Replace with `!is.empty(updatedFiles)`

File: lib/workers/repository/onboarding/branch/config.js [hunk 7]
```diff
@@ -25,7 +26,7 @@ async function getOnboardingConfig(config) {
         String(branchName).replace('greenkeeper', 'renovate'),
       ];
     }
-    if (Array.isArray(ignore) && ignore.length) {
+    if (is.array(ignore) && ignore.length) {
```
Comment (rarkins) [inline lib/workers/repository/onboarding/branch/config.js]: Replace with `!is.empty(ignore)`

**Review by rarkins — CHANGES_REQUESTED** [review]
Some is.empty() uses recommended

**Comment by rarkins** [comment]
Also please regenerate the `yarn.lock` and commit

File: lib/platform/github/index.js (hunk 5 above)
Comment (connorbrathwaite) [reply lib/platform/github/index.js]: this one caused a snapshot test to fail unfortunately, so for the time being i've left it as is

**Comment by connorbrathwaite** [comment]
updated with requested changes, let me know if theres anything else to improve/fix!

**Comment by rarkins** [comment]
CI still failing. @connorbrathwaite did you see the same failures locally when running `yarn test` ?

**Comment by connorbrathwaite** [comment]
huh, afaik the bump in jest versions was causing the auto mock in `test/workers/repository/init/index.spec.js` to raise an exception, merging with upstream/master took care of it though /shrug

**Comment by rarkins** [comment]
automock has caused me nothing but problems, maybe I should remove it. Unfortunately looks like it's back failing again though:

image

**Comment by rarkins** [comment]
automock remove from master and merged here - let's see how it goes

**Review by rarkins — APPROVED** [review]

**Comment by rarkins** [comment]
Passing tests now but somehow we have a coverage regression in lib/config/decrypt.js

**Comment by rarkins** [comment]
If you run `yarn jest` locally, does it report 100%? That command uses slight different reporting than codecov's.

**Comment by rarkins** [comment]
I prefer not to ignore them until I understand why the coverage changed. i.e. are the is.object statements returning something different? I will clone your fork and try to dig into it soon.

**Comment by rarkins** [comment]
@connorbrathwaite I found why the coverage went down - `is.object([])` returns `true` whereas for our internal function before it would be `false`. Hence the `is.array()` call after it is never getting reached. I need to code review it again to see if we're missing any more like that.

**Comment by rarkins** [comment]
Thanks @connorbrathwaite !

**Pull request merged**